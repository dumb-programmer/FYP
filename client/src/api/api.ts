export function login(data: { email: string; password: string }) {
  return fetch("http://localhost:3000/login", {
    method: "POST",
    body: JSON.stringify(data),
    headers: {
      "Content-Type": "application/json",
    },
    mode: "cors",
    credentials: "include",
  });
}

export function signup(data: {
  firstName: string;
  lastName: string;
  email: string;
  password: string;
}) {
  return fetch("http://localhost:3000/signup", {
    method: "POST",
    body: JSON.stringify(data),
    headers: {
      "Content-Type": "application/json",
    },
    mode: "cors",
  });
}

export function getMessages(chatId: string, page: number) {
  return fetch(`http://localhost:3000/chats/${chatId}/messages?page=${page}`, {
    mode: "cors",
    credentials: "include",
  });
}

export function searchMessages(chatId: string, query: string, page: number) {
  return fetch(
    `http://localhost:3000/chats/${chatId}/messages?query=${query}&page=${page}`,
    {
      mode: "cors",
      credentials: "include",
    }
  );
}

export function deleteMessage(chatId: string, messageId: string) {
  return fetch(`http://localhost:3000/chats/${chatId}/messages/${messageId}`, {
    method: "DELETE",
    mode: "cors",
    credentials: "include",
  });
}

export function sendPrompt(data: { prompt: string }, chatId?: string) {
  return fetch(`http://localhost:3000/chats/${chatId}`, {
    method: "POST",
    mode: "cors",
    credentials: "include",
    body: JSON.stringify(data),
    headers: {
      "Content-Type": "application/json",
    },
  });
}

export function getChats(page: number) {
  return fetch(`http://localhost:3000/chats?page=${page}`, {
    mode: "cors",
    credentials: "include",
  });
}

export function updateChat(chatId: string, name: string) {
  return fetch(`http://localhost:3000/chats/${chatId}`, {
    method: "PATCH",
    body: JSON.stringify({ name }),
    headers: {
      "Content-Type": "application/json",
    },
    credentials: "include",
    mode: "cors",
  });
}

export function deleteChat(chatId: string) {
  return fetch(`http://localhost:3000/chats/${chatId}`, {
    method: "DELETE",
    credentials: "include",
    mode: "cors",
  });
}
