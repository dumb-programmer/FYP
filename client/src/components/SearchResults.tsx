import MessageList from "./MessageList";
import { searchMessages } from "@/api/api";
import { useParams } from "react-router-dom";
import { useInfiniteQuery } from "react-query";
import InfiniteScrollContainer from "./InfiniteScrollContainer";

export default function SearchResults({ query }: { query: string }) {
    const { chatId } = useParams();
    const { data, isFetchingNextPage, hasNextPage, fetchNextPage, refetch } = useInfiniteQuery(`chat-${chatId}?${query}`, {
        queryFn: async ({ pageParam = 1 }) => {
            const response = await searchMessages(chatId, query, pageParam);
            if (response.ok) {
                return await response.json();
            }
        },
        getNextPageParam: (lastPage) => {
            if (lastPage.hasMore) {
                return lastPage.nextPage
            }
            return undefined
        }
    });

    return (
        <InfiniteScrollContainer className="flex-1 p-10 flex flex-col gap-10 overflow-y-auto" hasNextPage={hasNextPage} isFetchingNextPage={isFetchingNextPage} fetchNextPage={fetchNextPage}>
            {
                data?.pages && <MessageList refetchMessages={refetch} messages={[].concat(...data.pages.map(page => [...page.messages].reverse()).reverse())} />
            }
        </InfiniteScrollContainer>
    )
}
