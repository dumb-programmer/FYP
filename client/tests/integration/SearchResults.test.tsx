import React from "react";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { expect, describe, it, vi, beforeEach, afterEach } from "vitest";
import { QueryClient, QueryClientProvider } from "react-query";
import { MemoryRouter, Route, Routes } from "react-router-dom";
import SearchResults from "../../src/components/SearchResults";
import { searchMessages } from "../../src/api/api";

vi.mock("../../src/api/api", async () => ({
    ...await vi.importActual("../../src/api/api"),
    searchMessages: vi.fn()
}));

const queryClient = new QueryClient();

const renderComponent = (query: string, chatId: string) => {
    return render(
        <QueryClientProvider client={queryClient}>
            <MemoryRouter initialEntries={[`/chat/${chatId}`]}>
                <Routes>
                    <Route path="/chat/:chatId" element={<SearchResults query={query} />} />
                </Routes>
            </MemoryRouter>
        </QueryClientProvider>
    );
};

describe("SearchResults", () => {
    beforeEach(() => {
        window.IntersectionObserver = vi.fn(() => ({ observe: vi.fn(), disconnect: vi.fn() }))
        HTMLDivElement.prototype.scrollBy = vi.fn()
    });

    afterEach(() => {
        vi.clearAllMocks();
    });

    it("renders correctly with messages", async () => {
        const mockMessages = [{ _id: "1", prompt: "A", response: "B" }, { _id: "2", prompt: "C", response: "D" }];
        const mockResponse = {
            ok: true,
            json: async () => ({
                messages: mockMessages,
                hasMore: false,
                nextPage: undefined,
            }),
        };

        searchMessages.mockResolvedValue(mockResponse as any);

        const { container } = renderComponent("test", "123");

        expect(container).toMatchSnapshot();

        await waitFor(() => {
            expect(screen.getByText("A")).toBeInTheDocument();
            expect(screen.getByText("B")).toBeInTheDocument();
            expect(screen.getByText("C")).toBeInTheDocument();
            expect(screen.getByText("D")).toBeInTheDocument();

        });

    });

    it("renders empty state if no messages", async () => {
        const mockResponse = {
            ok: true,
            json: async () => ({
                messages: [],
                hasMore: false,
                nextPage: undefined,
            }),
        };

        searchMessages.mockResolvedValue(mockResponse as any);

        renderComponent("test", "123");

        await waitFor(() => {
            expect(screen.queryByText("A")).not.toBeInTheDocument();
            expect(screen.queryByText("B")).not.toBeInTheDocument();
            expect(screen.queryByText("C")).not.toBeInTheDocument();
            expect(screen.queryByText("D")).not.toBeInTheDocument();

        });
    });

    it("calls searchMessages with correct parameters", async () => {
        const mockResponse = {
            ok: true,
            json: async () => ({
                messages: [{ _id: "1", prompt: "Hello", response: "Hi" }],
                hasMore: false,
                nextPage: undefined,
            }),
        };

        searchMessages.mockResolvedValue(mockResponse as any);

        renderComponent("test", "123");

        await waitFor(() => {
            expect(searchMessages).toHaveBeenCalledWith("123", "test", 1);
        });
    });

    it("calls searchMessages with correct parameters, even if there are more pages", async () => {
        vi.mock("../../src/hooks/useIntersectionObserver", () => ({ default: () => ({ ref: vi.fn(), inView: true }) }))

        const mockResponsePage1 = {
            ok: true,
            json: async () => ({
                messages: [{ _id: "1", prompt: "Hello", response: "Hi" }],
                hasMore: true,
                nextPage: 2,
            }),
        };

        const mockResponsePage2 = {
            ok: true,
            json: async () => ({
                messages: [{
                    _id: "2", prompt: "A", response: "B"
                }],
                hasMore: false,
                nextPage: undefined,
            }),
        };

        searchMessages.mockImplementation((chatId, query, page) => {
            if (page === 1) return mockResponsePage1;
            return mockResponsePage2;
        })

        const { container } = renderComponent("test", "123");

        await waitFor(() => {
            expect(searchMessages).toHaveBeenCalledWith("123", "test", 1);
        });

        fireEvent.scroll(container.querySelector("div") as HTMLDivElement, { target: { scrollTop: 0 } })

        await waitFor(() => {
            expect(searchMessages).toHaveBeenCalledWith("123", "test", 2);
        });

    });
});
