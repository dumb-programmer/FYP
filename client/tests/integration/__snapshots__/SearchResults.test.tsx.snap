// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchResults > renders correctly with messages 1`] = `
<div>
  <div
    class="flex-1 p-10 flex flex-col gap-10 overflow-y-auto"
    id="messages-container"
  >
    <div />
  </div>
</div>
`;
